### PR TITLE
Add docker image with prebuilt 4c

### PR DESCRIPTION
**Pull request by ppraegla**
This PR adds a new docker image where 4C is already built. The plan is to create the docker image once the nightly workflow succeeds. Since we do not have versioning yet, the image is always tagged as `latest`. This docker image can be used in Actions where a 4C build is required or for users who want to quickly check out 4C without needing to build it themselves.

~~This is a first draft as I wanted to get your feedback first, before rebuilding the `4c-dependencies` image because I moved a Dockerfile.~~

The Dockerfile for the dependencies image has been moved to a separate folder in #112. The Dockerfile to build 4c is added to the new folder `prebuilt_4C`.

## Additional info
The pre-built 4C docker image just uses the `4c-dependencies-ubuntu24.04` image out of convenience for now. Originally, I wanted to use a multi-stage build where I just copy the executable to the final image. But then the tests did not work because they need the dat file from the source. I think as a start, we can have the source and build in the final docker image. Later we can focus how we can reduce the size of the docker image because 7.4 GB is huge.

@isteinbrecher  @gilrrei  @davidrudlstorfer  I think this image is a good starting point for further discussions for your use cases in Queens and for the actions meshpy.

## How to use the 4c docker image
Start the docker container with the current directory mounted to `/workspace`
```
docker run --interactive --tty --mount type=bind,source=./,target=/workspace ghcr.io/ppraegla/4c:latest
```
Set the correct owner of the mount (Unfortunately, we need to do that because we are the user `user` inside the docker image and the mount point is owned by root)
```
sudo chown user:user -R /workspace/
```
Run 4C with an input file from the host system and the output is also written to the host system
```
mpirun -np 2 /home/user/4C/build/4C /workspace/particle_dem_1d_radius_from_input.dat /workspace/out
```

This whole process is still a little awkward to use. But may be a good starting point to work with.

### CODE REVIEW

**Comment by sebproell** [comment]
> One could copy input files to the build directory during configuration. Then, the build and its tests are independent of the source location. Other projects do that.

@mayrmt Yes, we could do this. I normally configure all my necessary support files into the build tree to have a self-sufficient build folder after building. I think I already talked with @davidrudlstorfer about that in the context of #56.

@ppraegla The question would also be if we really need/want to configure and build all the tests in the docker image. If the image is only meant for users to get quick access to a pre-built 4C, the test suite is not going to be very interesting for them. This might depend on the envisioned use case, I just wanted to mention that we might not need to solve this problem at all.

File: docker/prebuilt_4C/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+FROM ghcr.io/4c-multiphysics/4c-dependencies-ubuntu24.04:latest
+LABEL org.opencontainers.image.description="Image containing the built 4C code"
+LABEL org.4c-multiphysics.project=4C
+
+RUN git clone https://github.com/4C-multiphysics/4C.git
+
+RUN mkdir 4C/build && cd 4C/build && \
+  cmake ../ --fresh --preset=docker && \
+  cmake --build . --target full -- -j `nproc`
```
Comment (sebproell) [inline docker/prebuilt_4C/Dockerfile:16]: Once #44 works, we could install it here and remove the build artifacts.

**Review by sebproell — COMMENTED** [review]
Nice! Amazing how little code and configuration this requires. Our work on the build system and CI infrastructure really pays off here.

**Comment by ppraegla** [comment]
@sebproell @mayrmt I also thought about what makes sense to include in the docker image (only the plain executable or also the source with the ability to run tests). Since we still have all the testing dependencies in the image, I just included the source as well (This was also the easiest solution). But I think it would make sense to create a bare minimum image with only the exectuable.

**Comment by sebproell** [comment]
@ppraegla I am also OK with the current state as a first working version. Feel free to build the image.

**Comment by isteinbrecher** [comment]
@ppraegla thanks a lot for doing this! I will have a look on how to integrate this into the MeshPy testing. This will hopefully save a lot of manual work in the future.

**Comment by isteinbrecher** [comment]
I just tried it myself, and I am amazed how easy it is to run 4C (compared to 6 years ago, when I first saw it). I am very much in favour of putting this high up on our "getting started with 4C" instructions. 🚀

**Comment by ppraegla** [comment]
The `workflow_run` will only work once it is on the default branch. But I tested the `workflow_run` [here](https://github.com/ppraegla/github-actions-playground/actions/runs/12175810683) which triggered [this](https://github.com/ppraegla/github-actions-playground/actions/runs/12175813877) workflow. So, it should be good to go :crossed_fingers:

File: docker/prebuilt_4C/Dockerfile (hunk 1 above)
Comment (ppraegla) [reply docker/prebuilt_4C/Dockerfile:16]: Good point. We need to clean up the docker image anyway once we know what exactly we want.

**Review by sebproell — APPROVED** [review]

**Review by isteinbrecher — APPROVED** [review]
Thanks a lot again!

File: .github/workflows/docker_prebuilt_4c.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          # tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+      # This step generates an artifact attestation for the image, which is an unforgeable statement about where and how it was built. It increases supply chain security for people who consume the image. For more information, see "[AUTOTITLE](/actions/security-guides/using-artifact-attestations-to-establish-provenance-for-builds)."
+      - name: Generate artifact attestation
+        uses: actions/attest-build-provenance@ef244123eb79f2f7a7e75d99086184180e6d0018 # v1.1.4
+        with:
+          subject-name: ${{ env.IMAGE_NAME }}
+          subject-digest: ${{ steps.push.outputs.digest }}
+          push-to-registry: true
```
Comment (amgebauer) [inline .github/workflows/docker_prebuilt_4c.yml:62]: Can we generalize the action `build_dependencies` that we already have? We could call it `build_and_publish_docker_image` (or so) and pass the base image, Dockerfile, imagename etc.

For dependencies, we pass `ubuntu:2{0,4}.04` and for this one here, we can pass our dependencies image.

**Review by amgebauer — COMMENTED** [review]
Nice work!

**Review by amgebauer — APPROVED** [review]

**Pull request merged**